src/Microsoft.ML.Core/Utilities/FuncMethodInfo3`3.cs

```diff
@@ -0,0 +1,48 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Immutable;
+using System.Reflection;
+using Microsoft.ML.Runtime;
+
+namespace Microsoft.ML.Internal.Utilities
+{
+    /// <summary>
+    /// Represents the <see cref="MethodInfo"/> for a generic function corresponding to <see cref="Func{T1, T2, TResult}"/>,
+    /// with the following characteristics:
+    ///
+    /// <list type="bullet">
+    /// <item><description>Three generic type arguments.</description></item>
+    /// <item><description>A return value of <typeparamref name="TResult"/>.</description></item>
+    /// </list>
+    /// </summary>
+    /// <typeparam name="T1">The type of the first parameter of the method.</typeparam>
+    /// <typeparam name="T2">The type of the second parameter of the method.</typeparam>
+    /// <typeparam name="TResult">The type of the return value of the method.</typeparam>
+    internal abstract class FuncMethodInfo3<T1, T2, TResult> : FuncMethodInfo<T1, T2, TResult>
+    {
+        private ImmutableDictionary<(Type, Type, Type), MethodInfo> _instanceMethodInfo;
+
```

Comment:
Some classes have this ImmutableDictionary. Some don't. Can you please help me understand the difference?

Reply:
In all cases, it's a cache for `MakeGenericMethod`